src/ir/operator.h

```diff
@@ -0,0 +1,54 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#ifndef SRC_IR_OPERATOR_H_
+#define SRC_IR_OPERATOR_H_
+
+#include <vector>
+
+#include "absl/container/flat_hash_map.h"
+#include "src/ir/types/type.h"
+#include "src/ir/value.h"
+
+namespace raksha::ir {
+
+class Block;
+
+// Signature of an operator. An operator could be simple operators (e.g., `+`,
+// `-`, `sql.select`, `sql.groupby`) or complex operators (e.g., particle,
+// function) that is compose of other operations.
+class Operator {
+ public:
+  using Name = std::string;
+
+  Operator(std::string name, std::unique_ptr<Block> implementation)
```

Comment:
If you're defining a type of `Name` to `std::string`, you should probably use this in the parameter.

One thing I'm wondering though: should this be a `struct` wrapper rather than just a type alias? A type alias will not prevent assigning to other kinds of `std::strings` without a method or ctor call. If you want the type system to protect you here, just defining a simple:

```C++
struct Name {
  std::string name_str;
}
```

would go a long way towards letting the type system protect you a bit. You could even use the new C++ designated initializer syntax and the public field to wrap and unwrap it quickly without having to define a bunch of boilerplate.

Reply:
Nah, if it doesn't provide immediate benefits don't worry about it. We can always add it later if we start having dataflow confusion on names.